### PR TITLE
feat(mongo/document/validation): Add distinction between internal and validation errors

### LINKED ISSUE #551: [mongo/document] Validate may return internal errors in addition of the validation errors
`ValidationErrors` are errors returned as-is to the end-user. However we may face internal errors as part of the validation process. At the moment we are not able to return such error to prevent saving the document without returning it as-is to the end user.

This issue is a suggestion to transform the following:
```go
func (sut Subject) Validate(ctx context.Context) *document.ValidationErrors {
	validationError := document.NewValidationErrorsBuilder()

	var sutWas Subject
	err := document.Find(ctx, SubjectCollection, sut.ID, &sutWas)
	if err != nil {
		log.WithError(err).Errorf("Failed to query for subject %s", sut.ID.Hex())
		validationError.Set("something", "internal server error")
	}

	return validationError.Build()
}
```

into the following:
```go
func (sut Subject) Validate(ctx context.Context) (*document.ValidationErrors, error) {
	validationError := document.NewValidationErrorsBuilder()

	var sutWas Subject
	err := document.Find(ctx, SubjectCollection, sut.ID, &sutWas)
	if err != nil {
		return nil, err
	}

	return validationError.Build(), nil
}
```

That way, `document.Save` would be able to make the difference between an internal error and a validation error, and thus react accordingly

### PR DESCRIPTION
**Pull request by aurelien-reeves-scalingo**
Fix #551 
- [x] Add a changelog entry in `CHANGELOG.md`

This PR would make possible such thing:
```go
		err := document.Save(ctx, collection, subject)
		if err != nil {
			validationErr := &pkgerrors.ValidationErrors{}
			if errors.As(pkgerrors.RootCause(err), &validationErr) {
				// something like showing it to the user
			} else {
				// something else like returning a 500
			}
		}
```

### CODE REVIEW

**Comment by aurelien-reeves-scalingo** [comment]
Can't remember why I did not finished that one at that time.

As this is a breaking change, due to my change of team, I would not have the opportunity to go further with this (releasing it, updating impacted projects, ...).

So feel free to close it

File: mongo/document/document.go [hunk 1]
```diff
@@ -46,22 +46,30 @@ type Closer interface {
 }
 
 type Validable interface {
-	Validate(ctx context.Context) *ValidationErrors
+	Validate(ctx context.Context) (*ValidationErrors, error)
```
[the remark this answers is not in the document]
Comment (aurelien-reeves-scalingo) [reply mongo/document/document.go:49]: How would you do that?

[the remark this answers is not in the document]
Comment (aurelien-reeves-scalingo) [reply mongo/document/document.go:49]: Oh, I remember.

Yes, we are able to make the distinction between a validation error and an internal error that would have happened during the save part.

But if an Internal error is raised as part of the validation process, we cannot

[the remark this answers is not in the document]
Comment (aurelien-reeves-scalingo) [reply mongo/document/document.go:49]: It is the case here: https://github.com/Scalingo/appsdeck-database/pull/2728/files/85e7636eafd0515fa05f5cd5b02c7433cda9be46#diff-fb3dafe5420d7f35c4af24c65542d1d8e1aa6cfedffa11d8ddca206e91503f2aR148

There is an internal error part of a `validate` method. It is not a validation error, but as it is part of the validate method, it is processed as a validation error.

[the remark this answers is not in the document]
Comment (EtienneM) [reply mongo/document/document.go:49]: Aaaaah yes I see the use case, I completely forgot about this. Thanks :)

**Review by EtienneM — APPROVED** [review]

**Review by EtienneM — CHANGES_REQUESTED** [review]
Oh no I can't validate, wait a sec :sweat_smile:

**Review by EtienneM — COMMENTED** [review]
`document.ValidationErrors` has been moved to `errors.ValidationErrors` but kept for retro-compatibility purpose.

If we release a new major version of this library, I think we should remove `document.ValidationErrors` and update the `Validate` signature to target `errors.ValidationErrors`. What do you think?

**Comment by aurelien-reeves-scalingo** [comment]
> `document.ValidationErrors` has been moved to `errors.ValidationErrors` but kept for retro-compatibility purpose.
> 
> If we release a new major version of this library, I think we should remove `document.ValidationErrors` and update the `Validate` signature to target `errors.ValidationErrors`. What do you think?

Good idea.

What about opening a PR over that one?

**Review by EtienneM — APPROVED** [review]

**Comment by aurelien-reeves-scalingo** [comment]
I did not realized how easy would be to remove the aliased type `document.ValidationErrors`.

@EtienneM could you please review it again?

**Review by EtienneM — APPROVED** [review]

**Pull request merged**